backend/app/synthesis/synthesizers.py

```diff
@@ -3,44 +3,103 @@
 """
 import math
 import numpy as np
-
+from scipy import signal
 
 from app.common import NOTE_FREQ_SIMPLE
 from app.synthesis.audio_encoding import WAV_SAMPLE_RATE
 
 
-def generate_sine_wave(frequency, duration):
+# pylint: disable=too-many-locals
+def generate_wave(frequency, duration, harmonics=0, vibrato=False, wave_type=np.sin):
+    """
+    Generates audio samples for various wave types
+    :param frequency: frequency in Hz
+    :param duration:  duration in seconds
+    :param harmonics: number of desired harmonics
+    :param wave_type:  wave type
+    :param vibrato: boolean for using vibrato or not
+    :return: audio samples for wave of the given wave type
+    """
+    num_samples = int(duration * WAV_SAMPLE_RATE)
+    time_steps = np.linspace(0, duration, num=num_samples, retstep=False)
+    wave_samples = wave_type(frequency * 2 * np.pi * time_steps)
+
+    if harmonics != 0:
+        for i in range(harmonics):
+            harmonic = np.sin(frequency * 2 * np.pi * time_steps * (i + 2))
+            for j, _ in enumerate(wave_samples):
+                wave_samples[j] += harmonic[j]
+
+    if vibrato:
```

Comment:
Since this vibrato procedure can function on a buffer of audio samples without needing any reference to the rest of the args for this function, I'd suggest factoring it out into its own function for clarity, and applying it afterwards, rather than during `generate_wave`.